File: app/src/main/java/b2w/com/alodjinha/Adapter/CategoriesAdapter.java
package b2w.com.alodjinha.Adapter;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.squareup.picasso.Picasso;

import java.util.List;

import b2w.com.alodjinha.Interface.ItemClickListener;
import b2w.com.alodjinha.Model.Category;
import b2w.com.alodjinha.R;
import b2w.com.alodjinha.ShowCategory;
import b2w.com.alodjinha.Util.Common;
import b2w.com.alodjinha.ViewHolder.ShowCategoryViewHolder;

public class CategoriesAdapter extends RecyclerView.Adapter<ShowCategoryViewHolder> {

    private Context context;
    private List<Category> listCategories;

    public CategoriesAdapter(Context context, List<Category> listCategories) {
        this.context = context;
        this.listCategories = listCategories;
    }

    @Override
    public ShowCategoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View itemView = LayoutInflater.from(context)
                .inflate( R.layout.layout_category, parent, false );

        return new ShowCategoryViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ShowCategoryViewHolder holder, final int position) {

        final Category category = listCategories.get(position);

        Picasso.with( context ).load( category.getUrlImage() ).into( holder.image_category );
        holder.name_category.setText( category.getDescription() );

        holder.setItemClickListener(new ItemClickListener() {
            @Override
            public void onClick(View view, int position, boolean isLongClick) {

                Intent intent = new Intent( context, ShowCategory.class );
                Common.selectedCategory = category;
                context.startActivity( intent );

            }
        });


    }

    @Override
    public int getItemCount() {
        return listCategories.size();
    }

    public void removeItem( int position ){
        listCategories.remove(position);
    }

    public void restoreItem( Category item, int position ){
        listCategories.add(position, item);
    }

    public Category getItem(int position){
        return listCategories.get(position);
    }

}

File: view/src/main/java/alodjinha/cfgdemelo/com/view/adapter/CategoriesAdapter.kt
package alodjinha.cfgdemelo.com.view.adapter

import alodjinha.cfgdemelo.com.model.Category
import alodjinha.cfgdemelo.com.view.R
import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso

class CategoriesAdapter(val context: Context, val categories: List<Category>, private val categoryClick: CategoryClickListener): RecyclerView.Adapter<CategoriesAdapter.RecyclerViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, p1: Int): CategoriesAdapter.RecyclerViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.category_view_holder, parent, false)

        return RecyclerViewHolder(view, categoryClick)
    }

    override fun getItemCount(): Int {
        return categories.count()
    }

    override fun onBindViewHolder(holder: CategoriesAdapter.RecyclerViewHolder, position: Int) {
        holder.bindData(context, categories, position)
    }

    class RecyclerViewHolder(itemView: View, private val categoryClick: CategoryClickListener): RecyclerView.ViewHolder(itemView) {
        private val ivCategory: ImageView = itemView.findViewById(R.id.ivCategory)
        private val tvCategory: TextView = itemView.findViewById(R.id.tvCategory)

        fun bindData(context: Context, categories: List<Category>, position: Int) {
            Picasso.with(context)
                .load(categories[position].imageUrl)
                .into(ivCategory)
            tvCategory.text = categories[position].description
            itemView.setOnClickListener {
                categoryClick.getCategory(context, categories[position])
            }
        }
    }

    interface CategoryClickListener {
        fun getCategory(context: Context, category: Category)
    }

}